server/raft_test.go

```diff
@@ -844,3 +844,55 @@ func TestNRGNoResetOnAppendEntryResponse(t *testing.T) {
 	require_NotEqual(t, leader.pterm, 0)
 	require_NotEqual(t, leader.pindex, 0)
 }
+
+func TestNRGCandidateDontStepdownDueToLeaderOfPreviousTerm(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+	c.waitOnLeader()
+
+	nc, _ := jsClientConnect(t, c.leader(), nats.UserInfo("admin", "s3cr3t!"))
+	defer nc.Close()
+
+	rg := c.createRaftGroup("TEST", 3, newStateAdder)
+	rg.waitOnLeader()
+
+	// create a candidate that has received entries while they were a follower in a previous term
+	candidate := rg.nonLeader().node().(*raft)
+	candidate.Lock()
+	candidate.switchState(Candidate)
+	candidate.pterm = 50
+	candidate.pindex = 70
+	candidate.term = 100
+	candidate.Unlock()
+
+	// leader term is behind candidate
+	leader := rg.leader().node().(*raft)
+	leader.Lock()
+	leader.term = candidate.pterm
+	leader.pterm = candidate.pterm
+	leader.pindex = candidate.pindex
+	leader.Unlock()
+
+	// Subscribe to the append entry subject.
+	sub, err := nc.SubscribeSync(leader.asubj)
+	require_NoError(t, err)
+
+	// Get the first append entry that we receive, should be heartbeat from leader of prev term
+	msg, err := sub.NextMsg(5 * time.Second)
+	require_NoError(t, err)
+
+	// Decode the append entry
+	ae, err := leader.decodeAppendEntry(msg.Data, nil, msg.Reply)
+	require_NoError(t, err)
+
+	// check that the append entry is from the leader of the previous term
+	require_Equal(t, ae.term, leader.term)
+	require_Equal(t, ae.leader, leader.id)
+
+	// check that we haven't stepped down
+	require_Equal(t, candidate.State(), Candidate)
+	// check that our term is still ahead of the leader's term
+	if candidate.term <= ae.term {
+		t.Fatalf("Expected candidate term to be ahead of leader term")
```

Comment:
I'd probably just do something like `require_True(t, candidate.term > ae.term)` here for brevity.